### DISCUSSION TITLE
How do I set a type when passing abi as a param to a function?

### DISCUSSION THREAD

**Discussion by benesjan** [wevm/viem#416] [Question]
Hello, I am try to create a function which deploys arbitrary contract and I am struggling with setting a type for abi.

<img width="670" alt="image" src="https://user-images.githubusercontent.com/13470840/234516508-d62c1035-236d-496d-af50-a373cbc311f8.png">

How would I do that?

Is there a type for "generic abi thing", "generic event thing" etc.?

**Comment by fubhy** [comment]
Yes there is. This imho deserves its own section in the docs at some point (how to utilize the type system and adding your own abstraction & functions around it). Best way to learn how to use the generics for this case is to look at the source code of the function(s)  you are trying to call. E.g. `deployContract` in this case: https://github.com/wagmi-dev/viem/blob/43867afa1f1368af3e69a3c6f7a4021a7af0a662/src/actions/wallet/deployContract.ts#L60-L88

What's most relevant here is the generic:

```ts
TAbi extends Abi | readonly unknown[]
```

For bytecode, btw., it's probably sufficient to just type it as `Hex` (there's also a Viem type for that which is simply an alias for `0x${string}`.

**Comment by fubhy** [comment]
Note that if you want to derive args from the abi you'll have to utilize other generics too (e.g. `GetFunctionArgs`, etc.)

**Comment by benesjan** [comment]
Using `Narrow<Abi | readonly unknown[]>` and `Hex` solved the issue.

```
async function deployL1Contract(
  walletClient: WalletClient<HttpTransport, Chain, Account>,
  publicClient: PublicClient<HttpTransport, Chain>,
  abi: Narrow<Abi | readonly unknown[]>,
  bytecode: Hex,
) {
  const hash = await walletClient.deployContract({
    abi,
    bytecode,
  });

  const receipt = await publicClient.waitForTransactionReceipt({ hash });
  const contractAddress = receipt.contractAddress;
  if (!contractAddress) {
    throw new Error(`No contract address found in receipt: ${JSON.stringify(receipt)}`);
  }

  return EthAddress.fromString(receipt.contractAddress!);
}
```

**Reply by tmm** [reply]
Worth noting that once we move the minimum TypeScript version to 5.0, you'll be able to drop `Narrow<...>` for the [`const` type parameter](https://www.totaltypescript.com/const-type-parameters).
